### PR TITLE
Better error message filter

### PR DESCRIPTION
**Pull request by doctrino**
## Description
Some user struggled to understand this error message:
image

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

**Comment by haakonvt** [comment]
So they passed in a set when the type says `Filter` or `dict`? Hmm......

**Comment by haakonvt** [comment]
Had a look again at your screenshot. This won't fix the user issue as the validate is called before "merge-space-into-filter" (which you have modified)

File: cognite/client/data_classes/filters.py [hunk 1]
```diff
@@ -238,9 +238,12 @@ def _validate_filter(
...
         return
-    if not_supported := (filter._involved_filter_types() - supported_filters):
-        names = [f.__name__ for f in not_supported]
-        raise ValueError(f"The filters {names} are not supported for {api_name}")
+    elif isinstance(filter, Filter):
+        if not_supported := (filter._involved_filter_types() - supported_filters):
+            names = [f.__name__ for f in not_supported]
+            raise ValueError(f"The filters {names} are not supported for {api_name}")
+    else:
+        raise ValueError(f"Invalid filter type: {type(filter)}. Must be Filter or dict")
```
Comment (haakonvt) [inline cognite/client/data_classes/filters.py:246]: This only raises if the top level filter is incorrect. Using e.g. `filter=And(MatchAll(), set())`, still gives the same error:
`AttributeError: 'set' object has no attribute '_involved_filter_types'`

Comment (doctrino) [reply cognite/client/data_classes/filters.py:249]: I think that is less likely user error to do, but added a check for that as well.

Comment (haakonvt) [reply cognite/client/data_classes/filters.py:249]: Nice

File: cognite/client/data_classes/filters.py [hunk 2]
```diff
@@ -193,7 +193,10 @@ def _involved_filter_types(self) -> set[type[Filter]]:
         output = {type(self)}
         if isinstance(self, CompoundFilter):
             for filter_ in self._filters:
-                output.update(filter_._involved_filter_types())
+                if isinstance(filter_, Filter):
+                    output.update(filter_._involved_filter_types())
+                else:
+                    raise ValueError(f"Unknown filter type: {type(filter_)}. Expected Filter got {type(filter_)}")
```
Comment (haakonvt) [inline cognite/client/data_classes/filters.py:199]: Should this rather be checked in `CompoundFilter.__init__`?
```suggestion
                    raise TypeError(f"Unknown filter type: {type(filter_)}")
```

File: cognite/client/data_classes/filters.py [hunk 3]
```diff
@@ -238,9 +241,12 @@ def _validate_filter(
...
         return
-    if not_supported := (filter._involved_filter_types() - supported_filters):
-        names = [f.__name__ for f in not_supported]
-        raise ValueError(f"The filters {names} are not supported for {api_name}")
+    elif isinstance(filter, Filter):
+        if not_supported := (filter._involved_filter_types() - supported_filters):
+            names = [f.__name__ for f in not_supported]
+            raise ValueError(f"The filters {names} are not supported for {api_name}")
+    else:
+        raise ValueError(f"Invalid filter type: {type(filter)}. Must be Filter or dict")
```
Comment (haakonvt) [inline cognite/client/data_classes/filters.py:249]:
```suggestion
    else:
        raise TypeError(f"Invalid filter type: {type(filter)}. Must be Filter or dict")
```

**Review by haakonvt — APPROVED** [review]
Pushed the changes we talked about 😄

**Pull request merged**